text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,483 @@
+### Last-value relationship
+
+Observer instruments have a well-defined _Last Value_ measured by the
+instrument, that can be useful in defining aggregations.  To maintain
+this property, we impose this requirement: two or more `Observe()`
+calls with an identical LabelSet during a single Observer callback
+invocation are treated as duplicates of each other, where the last
+call to `Observe()` wins.  (This is also the specified way that
+`Labels()` treats duplicates for a given label key.)
+
+Unlike Observer instruments, Measure instruments do not define the
+Last Value relationship.  One reason is that [synchronous events can
+happen
+simultaneously](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics.md#time).
+The Last Value relationship refers to values read in a single
+collection period, whereas with Measure instruments we can define a
+last-value aggregation.
+
+### Aggregating changes to a sum: Rate calculation
+
+The former `Monotonic` option had been introduced in order to support
+reporting of a current sum, such that a rate calculation is implied.
+Here we defined _Rate_ as an aggregation, defined for a subset of
+instruments, that may be calculated differently depending on how the
+instrument is defined.  The rate aggregation outputs the amount of
+change in a quantity divided by the amount of change in time.
+
+A rate can be computed from values that are reported as differences,
+referred to as _delta_ reporting here, or as sums, referred to as
+_cumulative_ reporting here.  The primary goal of the instrument
+refinements introduced in this proposal is to facilitate rate
+calculations in more than one way.
+
+When delta reporting, a rate is calculated by summing individual
+measurements or observations.  For Measure instruments, these values
+fall into a range of time, as indicated by the event timestamp.  For
+Observer instruments, these values fall into a range of collection
+intervals.
+
+When cumulative reporting, a rate is calculated by computing a
+difference between individual values.  For an Observer instrument, we
+compute rate over a range of collection intervals, and for a Measure
+instrument we compute rate over a range of timestamps.  In either
+case, we are interested in subtracting the final value from the prior
+value measured or observed on the instrument.
+
+Note that rate aggregation, as illustrated above, treats the time
+dimension differently than the other dimensions used for aggregation.
+
+### Standard implementation of Measure and Observer
+
+OpenTelemetry specifies how the default SDK should treat metric
+events, by default, when asked to export data from an instrument.
+Measure and Observer instruments compute `Sum` and `Count`
+aggregations, by default, in the standard implementation.  This pair
+of measurements, of course, defines an average value.  There are no
+restrictions placed on the numerical value in an event for the two
+foundational instruments.
```

Comment:
Just a thought: Not sure if we should ask for 2 values as default, because this implies a mutex in languages like java where it is hard to do atomically both operations. Maybe we can stick with only Sum as default for performance reasons.

Reply:
We have a lock-free solution to this in Go, I figure you could do that in Java if you wanted.
See https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/internal/state_locker.go or https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/metric/aggregator/minmaxsumcount/mmsc.go

Reply:
See https://github.com/open-telemetry/opentelemetry-go/pull/600 to prove my point.

Reply:
👨‍🔬